### PR TITLE
Base64 decoder rust/v4

### PR DESCRIPTION
**Pull request by inashivb**
Link to tickets:
- https://redmine.openinfosecfoundation.org/issues/7065
- https://redmine.openinfosecfoundation.org/issues/7058
- https://redmine.openinfosecfoundation.org/issues/6280

Previous PR: https://github.com/OISF/suricata/pull/11479

Changes since v3:
- fixed buffer overflow due to size constraints
- fixed CString conversion issues, length calculations

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 21504

**Comment by jasonish** [comment]
> @jasonish could you please take look at the rust code in this PR? Specially if what's done in [7ffcbad](https://github.com/OISF/suricata/pull/11487/commits/7ffcbad23218606f0368808afbc7d23d465d8567) is correct. Thank you! 🙇🏽‍♀️

It looks ok, but seems out of place?  The other conversion function in `log.rs` is there as its only used by that module, and is private to that module.  This utility function does seem out of place here though.

Some extra thoughts:
- In a function named `rs_base64_decode`, I am decoding base64. If that base64 has NULs in it, I want them. This looks like a specialized base64 decode that returns a sane C string. Perhaps the naming a function documentation should provide that..  Also, this is C name space...  Perhaps `SCBase64DecodeToCString`?
- The file can just be named `base64.rs`, it doesn't need the `sc`.

File: rust/src/log.rs [hunk 1]
```diff
@@ -202,6 +202,22 @@ pub fn sc_log_message(level: Level,
     return 0;
 }
 
+// Convert a Vec into a CString by first stripping NUL bytes.
+pub fn vec_to_safe_cstring(val: Vec<u8>) -> CString {
```
Comment (jasonish) [inline rust/src/log.rs:206]: As mentioned, seems out of place is the module. As it is only to support the specialized base64decode function, should probably live alongside it.

File: rust/src/utils/scbase64.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            }
+            decoder.nb = 0;
+        }
+    }
+    r.truncate(offset);
+    return Ok(r);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_base64_decode(input: *const u8, len: u32, max_decoded: u32, mode: Base64Mode) -> *mut Base64Decoded {
```
Comment (jasonish) [inline rust/src/utils/scbase64.rs:273]: As metioned, naming. Should follow `SCBase64Decode` format.. But `ToCString` added or something.

**Comment by jasonish** [comment]
One last thing to consider, after functionality is verified is merging with the existing `ffi:base64` module we have.

- The one under FFI is to just expose base64 functionality to C.
- The `utils/base64` could be nice pure Rust functions that provide all the functionality needed to Rust, which the functions in `ffi/rust` expose to C.

**Review by catenacyber — CHANGES_REQUESTED** [review]
Jason requested changes, and I think you are working on them already ;-)

**Pull request closed without merging**